src/measure_theory/pi_system.lean

```diff
@@ -268,9 +268,134 @@ begin
+  let g := λ n, (ite (n ∈ p1) (f1 n) set.univ) ∩ (ite (n ∈ p2) (f2 n) set.univ),
+  use [p1 ∪ p2, h_sup p1 p2 hp1S hp2S, g],
+  have h_inter_eq : t1 ∩ t2 = ⋂ i ∈ p1 ∪ p2, g i,
+  { rw [ht1_eq, ht2_eq],
+    simp_rw [← set.inf_eq_inter, g],
+    ext1 x,
+    simp only [inf_eq_inter, mem_inter_eq, mem_Inter, finset.mem_union],
+    refine ⟨λ h i hi_mem_union, _, λ h, ⟨λ i hi1, _, λ i hi2, _⟩⟩,
+    { split_ifs,
+      exacts [⟨h.1 i h_1, h.2 i h_2⟩, ⟨h.1 i h_1, set.mem_univ _⟩,
+        ⟨set.mem_univ _, h.2 i h_2⟩, ⟨set.mem_univ _, set.mem_univ _⟩], },
+    { specialize h i (or.inl hi1),
+      rw if_pos hi1 at h,
+      exact h.1, },
+    { specialize h i (or.inr hi2),
+      rw if_pos hi2 at h,
+      exact h.2, }, },
+  refine ⟨λ n hn, _, h_inter_eq⟩,
+  simp_rw g,
+  split_ifs with hn1 hn2,
+  { refine hpi n (f1 n) (hf1m n hn1) (f2 n) (hf2m n hn2) (set.ne_empty_iff_nonempty.mp (λ h, _)),
+    rw h_inter_eq at h_nonempty,
+    suffices h_empty : (⋂ i ∈ p1 ∪ p2, g i) = ∅,
+      from (set.not_nonempty_iff_eq_empty.mpr h_empty) h_nonempty,
+    refine le_antisymm (set.Inter_subset_of_subset n _) (set.empty_subset _),
+    refine set.Inter_subset_of_subset hn _,
+    simp_rw [g, if_pos hn1, if_pos hn2],
+    exact h.subset, },
+  { simp [hf1m n hn1], },
+  { simp [hf2m n h], },
+  { exact absurd hn (by simp [hn1, h]), },
+end
+
+lemma generate_from_pi_Union_Inter_le {α ι} {m : measurable_space α}
+  {s : ι → measurable_space α} (h : ∀ n, s n ≤ m) (π : ι → set (set α)) (S : set (finset ι))
+  (hpis : ∀ n, s n = generate_from (π n)) :
+  generate_from (pi_Union_Inter π S) ≤ m :=
+begin
+  refine generate_from_le _,
+  rintros t ⟨ht_p, ht_p_mem, ft, hft_mem_pi, rfl⟩,
+  refine finset.measurable_set_bInter _ (λ x hx_mem, (h x) _ _),
+  rw hpis x,
+  exact measurable_set_generate_from (hft_mem_pi x hx_mem),
+end
+
+lemma subset_pi_Union_Inter {α ι} {π : ι → set (set α)} {S : set (finset ι)}
+  (h_univ : ∀ i, set.univ ∈ (π i)) {i : ι} {s : finset ι} (hsS : s ∈ S) (his : i ∈ s) :
+  π i ⊆ pi_Union_Inter π S :=
+begin
+  refine λ t ht_pii, ⟨s, hsS, (λ j, ite (j = i) t set.univ), ⟨λ m h_pm, _, _⟩⟩,
+  { split_ifs,
+    { rwa h, },
+    { exact h_univ m, }, },
+  { ext1 x,
+    simp_rw set.mem_Inter,
+    split; intro hx,
+    { intros j h_p_j,
+      split_ifs,
+      { exact hx, },
+      { exact set.mem_univ _, }, },
+    { simpa using hx i his, }, },
+end
+
+lemma mem_pi_Union_Inter_of_measurable_set {α ι} (m : ι → measurable_space α)
+  {S : set (finset ι)} {i : ι} {t : finset ι} (htS : t ∈ S) (hit : i ∈ t) (s : set α)
+  (hs : measurable_set[m i] s) :
+  s ∈ pi_Union_Inter (λ n, {s | measurable_set[m n] s}) S :=
+subset_pi_Union_Inter (λ i, measurable_set.univ) htS hit hs
+
+lemma le_generate_from_pi_Union_Inter {α ι} {m : measurable_space α}
+  {π : ι → set (set α)} (S : set (finset ι)) (h_univ : ∀ n, set.univ ∈ (π n)) {x : ι}
+  {t : finset ι} (htS : t ∈ S) (hxt : x ∈ t) (hpix : m = measurable_space.generate_from (π x)) :
```

Comment:
again, I wouldn't use `m`, and express the conclusion as `generate_from (π x) ≤ generate_from (pi_Union_Inter π S)`.